### DISCUSSION TITLE
Multi language training

### DISCUSSION THREAD

**Discussion by Bebaam** [coqui-ai/TTS#2214] [General Q&A]
Hey everybody,

I currently try to train a multispeaker multilanguage model with phonemes. I've read before that I can start with one language and add new languages over time by training explicitly with the corresponding dataset (Searched a lot but unfortunately I can't find that thread, only https://github.com/coqui-ai/TTS/discussions/1859#discussioncomment-3396731). This would be perfect for me, because I can then set the proper phoneme language and a language dependent cleaner. 

However, I trained the model for close to 1 Million steps in German - with Thorsten voice and another one - where the quality is ok. But when I then want to add English with the LJ-Speech Dataset, using the English espeak phonemizer and an English cleaner, the capability to speak german and also the german speakers in general are lost directly (no difference in using each of the german speakers provided as 'speaker_idx' for inference). Do I need to manually add this information to force the model to keep speaker/language information?

As a possible alternative, if the method above is not working, maybe it is better to train with the whole dataset - containing German and English? As erogol pointed out in https://github.com/coqui-ai/TTS/discussions/1590#discussioncomment-2853891, phonemizer would be ready for that. But then I won't be able to use the cleaner appropriately? 

Thanks in advance for any help and insights.

**Comment by Ca-ressemble-a-du-fake** [comment]
Maybe you can use the "multilingual_cleaners" and use your multi language dataset ?
Also did you check the recipe they've just released [for multilingual training](https://github.com/coqui-ai/TTS/blob/061ac43187fbcc98ee73e158f672b7fea1e8da47/recipes/vctk/yourtts/train_yourtts.py) ?

**Reply by Bebaam** [reply]
Thanks for the hint with the new recipe, I'll try to adapt that. 
I remember that I've read somewhere that language specific cleaner is not neccessary when using espeak (as it supports both En and De). If that would be the case, using multilingual cleaners would be completely sufficient for me. 

Furthermore, I've just started to test using the whole dataset and with multilingual cleaners. The speaker information is still there, but quality directly drops. Maybe I have to train that for a bit. I'll try that in combination with information from the new recipe :)

**Reply by Bebaam** [reply]
I've tried that now, but the problem still exists at least after a few steps.
To get an idea, here are two samples saying the german sentence: "Heute ist ein großer Tag für uns alle."
The first one is generated with the german only model:

https://user-images.githubusercontent.com/44262699/207384859-98e9f7f9-ed5e-4a83-9f88-a0be64043396.mp4

 The second one is generated with the 2-language model after around 1000 steps:


https://user-images.githubusercontent.com/44262699/207385071-7d4190e3-1f7c-487b-9c2a-86247190d4a5.mp4

Something is going wrong, maybe it is a problem with the phonemes?

**Reply by Ca-ressemble-a-du-fake** [reply]
When I fine tune a model I train for more than 1000 steps. Maybe you've got to train 20k or 30k more steps and see. In tensorboard you can also have a look at the eval losses that should decrease (although I could never achieve to make them decrease, but he quality increased).

**Reply by Bebaam** [reply]
Yeah will do that definitively. Just wanted to give an insight. For me it sounded as something was wrong, hopefully it will diminish later on. 

Tensorboard is never really helpful for me, and I just listen to samples every few 10k steps, a bit like Edresson suggests here https://github.com/coqui-ai/TTS/discussions/1306#discussioncomment-2290489. So I think it is the same as for you, losses aren't that interesting, but indeed quality increases :)

**Comment by Ca-ressemble-a-du-fake** [comment]
Looking forward to hearing how it'll evolve overtime when you reach mote steps.

By the way the answer from Edresson you linked is very interesting : I've always been wondering how all those losses should be interpreted !

And you wrote that you trained for nearly 1M steps. How long did it take you to do that and what's your hardware / batch size ? With an i5 2400 (~10 years old) / 16GB / rtx 3090 I reach ~ 80k steps / 24h (bs = 32) so I'd have to wait ~13 days to hopefully reach the same quality than what you've got (dataset is mine with ~ 1500 samples).

**Reply by Bebaam** [reply]
I'll update you. My hardware is very similar. Ryzen 5600x / 32GB / RTX 3090. I also train with bs 32 and allow 1-13 seconds long audios. My training time is around 50k steps a day. I use wsl2 on windows, I should check using pure linux if there is such a huge difference in training time. 

Maybe I was already overfitting a bit, as my best checkpoint (I only checked randomly each few 10k steps) is around 840k, this is also the one I'll now use for the multilingual training.

**Reply by Bebaam** [reply]
My dataset for German is thorsten-de with around 22k samples, one with ~2.5k samples and lj speech with 13k.

**Reply by Bebaam** [reply]
40k steps later nothing changed :( Results only get worse.

**Reply by Ca-ressemble-a-du-fake** [reply]
One day I was told to freeze the duration predictor and the PE (post encoder I guess) when finetuning. I could never achieve better results with that but my datasets have always been reduced. So maybe it can work for you.

**Reply by Bebaam** [reply]
In which way you had your datasets reduced? 
With freezing I think you refer to https://github.com/coqui-ai/TTS/discussions/1781#discussioncomment-3230302?
Yeah I did not try freezing but hoped that at least decent results should be achieved without freezing.

**Reply by Ca-ressemble-a-du-fake** [reply]
I meant that my datasets have a reduced number of samples. So I use small datasets (~1500 samples) compared to what you use ! Listening to your eng + ger generated samples I would say that it lost something so yeah freezing should prevent this effect.

**Reply by Bebaam** [reply]
Ok I understand, maybe this can really have an effect here.

**Comment by Ca-ressemble-a-du-fake** [comment]
FWIW on my machine it's rather 90k per 24h as I've just measured it accurately on Tensorboard. So yeah you've got room for improvements! 
The first audio you provided sounds perfect to me. So what makes you think you were already overfitting a bit (I would love to be able to hear when the training has to be stopped because overfitting starts to occur)?

**Reply by Bebaam** [reply]
It was not an educated guess with the overfitting, in my opinion it just sounds a bit worse. 
Hard to describe, like unnatural pronounciation :D It mainly occured for the small dataset, not for the big Thorsten dataset (which apparently indicates that I should have used weighted speaker sampling like in the new recipe). 
Here are samples from step 841k and 900k for the small datset (Angela Merkel):

https://user-images.githubusercontent.com/44262699/207599843-cd471dcd-f1b7-4944-9c80-268745f1ea29.mp4

https://user-images.githubusercontent.com/44262699/207599873-609b7211-9f04-4d72-949f-f7b14d51fef0.mp4

Btw I train with sampling rate 22k.

**Reply by Ca-ressemble-a-du-fake** [reply]
I always train in 16kHz but as long as you keep the same sampling rate in all datasets it should be ok to use 22kHz. Did you use VITS ?

**Reply by Ca-ressemble-a-du-fake** [reply]
Oh and I remember that even Thorsten made a survey because he could not tell which of his trained models was the best one ! For me the first one (841k) was better because it sounds clearer.  But then I heard the second one again and liked it more because it was "sharper" and "stronger" (in layman terms😉).

**Reply by Bebaam** [reply]
> I always train in 16kHz but as long as you keep the same sampling rate in all datasets it should be ok to use 22kHz. Did you use VITS ?

Vits? I made a config file from the recipe/multilingual/vits_tts recipe. Do you mean that? 

For me the difference between 16kHz and 22kHz is enormous. And fortunately Thorsten an LjSpeech are both in 22kHz.

**Reply by Bebaam** [reply]
Yeah the sound quality is really subjective :D For me the second one just sounds less natural "German" :laughing: 

Btw I installed pure linux. I expected to have a loss of 10-20% which would be okay for me. But now training is up to 3x faster! It'll be around 130k now, thank you for that :)

**Reply by Ca-ressemble-a-du-fake** [reply]
🍾3x faster !  That's awesome ! It also shows in practice that decent cpu/motherboard is required to get all the high performance out of an rtx 3090 (your platform is 45% faster than mine for the same gpu and you're using 22k sampling rate so I ought to find the budget to upgrade my cpu😄). Regarding the voice quality  I don't know for sure whether it applies to your case or not but somebody told me that the quality of a single speaker model was better than the one of multispeaker model because [all neurons are dedicated to the voice] (https://gitter.im/coqui-ai/TTS?at=634ac1abcf41c67a5cb157c4). So that could partially explain the decrease in quality that you're experiencing.

**Reply by Bebaam** [reply]
Maybe there are some other factors influencing the speed, 45% seems a bit high. E.g. things like using an SSD (In RTVC repo it was sometimes of interest). Would be interesting to know where these huge differences exactly come from.

**Reply by Ca-ressemble-a-du-fake** [reply]
It has a decent SSD (from Samsung). I'd say that it comes from the transfer rate from GPU to mainboard components (CPU, RAM, mass storage...). But it's OK I am not in a hurry !
And yes I meant VITS model like you are using.

**Reply by Bebaam** [reply]
Ok yes that will be the case then, fortunately training is fast anyway :)

**Comment by p0p4k** [comment]
Can you incrementally add languages or speakers without losing the older languages?

**Reply by Bebaam** [reply]
Yes and no. When I train in German (for 1M steps) and add a second language afterwards (English) and train on the datasets of both languages together, the way the speaker pronounce the words in German is directly getting worse. I have the sample for the German only model (same as in answer above) for the sentence: "Heute ist ein großer Tag für uns alle.":

https://user-images.githubusercontent.com/44262699/207646501-7a853ebf-de91-4e4f-9845-efd50641b6a9.mp4

The model trained with German+English dataset after 1k steps:

https://user-images.githubusercontent.com/44262699/207645855-2b171acd-cdec-4f0a-bbb9-e8d4cdcde4f5.mp4

The model trained with German+English dataset after 40k steps:

https://user-images.githubusercontent.com/44262699/207645925-e3685086-fb0b-494e-9dd4-7b29bab6ab20.mp4

**Comment by Bebaam** [comment]
No idea what I did wrong, but now it is working. I just used another checkpoint and then recomputed the phoneme cache dir. Now the german language output is not corrupted anymore :) Thanks again @Ca-ressemble-a-du-fake for the support.

**Reply by Ca-ressemble-a-du-fake** [reply]
Cool:smile:! It hit me too when I was using a common folder for the phoneme cache. Now I append the cache dir to the current model dir in the recipe I use (so I only use the python way, not the json file anymore :relieved: )

**Reply by Bebaam** [reply]
Yes is probably the best way :smile:

**Comment by thivux** [comment]
hi @Bebaam, did you train with both German and English data or just continue training German data from English checkpoint?
